[operator/nodejs] document disabling node auto-instrumentations (#4824)
open-telemetry/opentelemetry.io: content/en/docs/kubernetes/operator/automatic.md

        value: http,nestjs-core # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

To keep all default libraries and disable only specific instrumentation
libraries you can use the `OTEL_NODE_DISABLED_INSTRUMENTATIONS` environment

> variable as documented in the
> [Node.js zero-code instrumentation documentation](/docs/zero-code/js/configuration/#excluding-instrumentation-libraries).

Reviewer: ```suggestion
variable. For details, see 
[Excluding instrumentation libraries](/docs/zero-code/js/configuration/#excluding-instrumentation-libraries).
```

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
# ... other fields skipped from this example
spec:
  # ... other fields skipped from this example
  nodejs:
    env:
      - name: OTEL_NODE_DISABLED_INSTRUMENTATIONS

>         value: fs # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.

Reviewer 2: Lets add a second package to show the comma-separated list in action.

Author: Added grpc for new example `fs,grpc`

        value: http,nestjs-core # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

To keep all default libraries and disable only specific instrumentation
libraries you can use the `OTEL_NODE_DISABLED_INSTRUMENTATIONS` environment
variable. For details, see
[Excluding instrumentation libraries](/docs/zero-code/js/configuration/#excluding-instrumentation-libraries).

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
# ... other fields skipped from this example
spec:
  # ... other fields skipped from this example
  nodejs:
    env:
      - name: OTEL_NODE_DISABLED_INSTRUMENTATIONS
        value: fs,grpc # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

> Note that `OTEL_NODE_ENABLED_INSTRUMENTATIONS` and
> `OTEL_NODE_DISABLED_INSTRUMENTATIONS` are mutually exclusive. If both are set,

> > all instrumentations will be disabled.

Reviewer 3: ```suggestion
> all instrumentations are disabled.
```

Author: Ah I pushed a commit changing the note type so it didn't let me commit this suggestion directly 😅  Made this change though, thanks for the explanation!

        value: http,nestjs-core # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

To keep all default libraries and disable only specific instrumentation
libraries you can use the `OTEL_NODE_DISABLED_INSTRUMENTATIONS` environment
variable. For details, see
[Excluding instrumentation libraries](/docs/zero-code/js/configuration/#excluding-instrumentation-libraries).

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
# ... other fields skipped from this example
spec:
  # ... other fields skipped from this example
  nodejs:
    env:
      - name: OTEL_NODE_DISABLED_INSTRUMENTATIONS
        value: fs,grpc # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

> > If both environment variables are set, `OTEL_NODE_ENABLED_INSTRUMENTATIONS` is
> > applied first, and then `OTEL_NODE_DISABLED_INSTRUMENTATIONS` is applied to
> > that list. Therefore, if the same instrumentation is included in both lists,
> > that instrumentation will be disabled.

Reviewer 4: ```suggestion
{{% alert title="Note" color="info" %}}

If both environment variables are set, `OTEL_NODE_ENABLED_INSTRUMENTATIONS` is
applied first, and then `OTEL_NODE_DISABLED_INSTRUMENTATIONS` is applied to
that list. Therefore, if the same instrumentation is included in both lists,
that instrumentation will be disabled.

{{% /alert %}}
```

        value: http,nestjs-core # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

To keep all default libraries and disable only specific instrumentation
libraries you can use the `OTEL_NODE_DISABLED_INSTRUMENTATIONS` environment
variable. For details, see
[Excluding instrumentation libraries](/docs/zero-code/js/configuration/#excluding-instrumentation-libraries).

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
# ... other fields skipped from this example
spec:
  # ... other fields skipped from this example
  nodejs:
    env:
      - name: OTEL_NODE_DISABLED_INSTRUMENTATIONS
        value: fs,grpc # comma-separated list of the instrumentation package names without the `@opentelemetry/instrumentation-` prefix.
```

{{% alert title="Notes" color="info" %}}

> Note that `OTEL_NODE_ENABLED_INSTRUMENTATIONS` and
> `OTEL_NODE_DISABLED_INSTRUMENTATIONS` are mutually exclusive. If both are set,
> all instrumentations are disabled.

Reviewer 5: Are you sure this is correct? My understanding is that if both are set, then first `*_ENABLED_*` is used to establish a starting set, then any of those listed in `*_DISABLED_*` are removed.

```
% cd ~/src/opentelemetry-js-contrib/metapackages/auto-instrumentations-node
% node
> var mod = require('./')
undefined
> process.env.OTEL_NODE_ENABLED_INSTRUMENTATIONS = 'dns,fs,hapi'
'dns,fs,hapi'
> mod.getNodeAutoInstrumentations().map(instr => instr.instrumentationName)
[
  '@opentelemetry/instrumentation-dns',
  '@opentelemetry/instrumentation-fs',
  '@opentelemetry/instrumentation-hapi'
]
> process.env.OTEL_NODE_DISABLED_INSTRUMENTATIONS = 'bunyan,fs'
'bunyan,fs'
> mod.getNodeAutoInstrumentations().map(instr => instr.instrumentationName)
[
  '@opentelemetry/instrumentation-dns',
  '@opentelemetry/instrumentation-hapi'
]
```

Perhaps:

```suggestion
If both environment variables are specified, `OTEL_NODE_ENABLED_INSTRUMENTATIONS` is first used to establish a set of instrumentation package names, then any names from `OTEL_NODE_DISABLED_INSTRUMENTATIONS` are removed from that set. Any remaining instrumentation names are enabled.
```

Author: Yeah no you're totally correct, I'm not sure what I was doing when I wrote the original test. Will update there and then here

Author: Updated!
